### ISSUE TITLE
ASP.NET Implicitly Enables Authorization and Authentication

### ISSUE THREAD

**Issue by BearsPunch** [dotnet/aspnetcore#48469]
I was interested to check how Authentication and Authorization works in ASP.NET so started debugging and was so surprised that even if I leave only 
```
var builder = WebApplication.CreateBuilder(args);
builder.Services.AddControllers();
var app = builder.Build();
app.MapControllers();
app.Run();
```
**AuthorizationMiddleware** and **AuthenticationMiddleware** still exist in request lifecycle. So, after that I found why it's happening:
1) **builder.Services.AddControllers** inside calls **AddAuthorization** which calls **services.AddAuthenticationCore**. It's already interesting why all those methods were called, but ok, it's just some dependency injection.
2) But after that **builder.Build** and class **WebApplicationBuilder** do next
```
 if (serviceProviderIsService?.IsService(typeof(IAuthenticationSchemeProvider)) is true)
        {
            // Don't add more than one instance of the middleware
            if (!_builtApplication.Properties.ContainsKey(AuthenticationMiddlewareSetKey))
            {
                // The Use invocations will set the property on the outer pipeline,
                // but we want to set it on the inner pipeline as well.
                _builtApplication.Properties[AuthenticationMiddlewareSetKey] = true;
                app.UseAuthentication();
            }
        }

        if (serviceProviderIsService?.IsService(typeof(IAuthorizationHandlerProvider)) is true)
        {
            if (!_builtApplication.Properties.ContainsKey(AuthorizationMiddlewareSetKey))
            {
                _builtApplication.Properties[AuthorizationMiddlewareSetKey] = true;
                app.UseAuthorization();
            }
        }
```
So, after that any app doesn't care if you set **app.UseAuthentication** and **app.UseAuthorization** or not, it always be here. Only quick way to fully disable authentication and authorization middleware is don't use **builder.Services.AddControllers**. 

Also noticed that asp.net as default will add **BlazorWasmHotReloadMiddleware** even if you don't use Blazor at all. Even if you don't use anything! 

So, ASP.NET loads many things what we don't need, and it easily can be added by developer via add/use methods.

It's just my quick view, maybe someone can me explain why controller can't work without auth systems. 
In Microsoft guide [link](https://learn.microsoft.com/en-us/aspnet/core/security/authentication/identity?view=aspnetcore-7.0&tabs=visual-studio) they use **builder.Services.AddRazorPages** with **app.UseAuthentication** and **app.UseAuthorization** but it also has no sense, because inside **AddRazorPages** method is the same what **AddControllers** method has.

**Comment by davidfowl** [comment]
Are you worried about the performance overhead?

**Comment by BearsPunch** [comment]
> Are you worried about the performance overhead?

I'm worried that the code is not doing what I expect. Also, I can replace **AddControllers** to **AddRazorPages** and magic, controllers still work and authentication with authorization also. I'm not a "click" guy, who needs just in two clicks make web app. I can agree if Web app template will have a lot of predefined services, which I can remove.

Just as example - I want to test app without any auth system (once it happened to me) So instead of removing all **authorize** attributes or set **AllowAnonymous**, I can just remove  **app.UseAuthentication** and **app.UseAuthorization**, but it will not help. Even if I remove everything what I see that could be related with auth, it still will not help. Every [authorize] attribute will still prevent access to controller.

**Comment by davidfowl** [comment]
Did you run into an issue? I'm not understanding the problem.

**Comment by BearsPunch** [comment]
> Did you run into an issue? I'm not understanding the problem.

I created issue about design. For me it's issue with design/method naming and Microsoft guidelines. If you expected **NullPointerException** somewhere, no, I don't have it. I did run into an issue when tried to disable authorization temporary. Also, I found that **AddMvcCore** didn't add any authorization, but again, by Microsoft guides **AddMvc** is old and I have to use **AddControllers** or **AddRazorPages**, but no info about **AddMvcCore** and no new **AddControllersCore** like methods.

**Comment by divyeshio** [comment]
This was surprising to me as well, why does framework adds authorization without explicitly asking for it? AddControllers should do what it says just add the controllers, Why do I need to pay for services that I didn't ask for. I do understand that to make building applications with asp.net more approachable for beginners, you need to hide certain complexity. But this is some unexpected addition of functionality from framework itself.

**Comment by davidfowl** [comment]
> 
This was surprising to me as well, why does framework adds authorization without explicitly asking for it? AddControllers should do what it says just add the controllers, Why do I need to pay for services that I didn't ask for. I do understand that to make building applications with asp.net more approachable for beginners, you need to hide certain complexity. But this is some unexpected addition of functionality from framework itself.

Did you run into beginner unfriendly behavior?

**Comment by divyeshio** [comment]
> Did you run into any unexpected behavior?

Nope, but then naming AddAuthentication and AddAuthorization is confusing as it is not adding anything new, but just used for configuring options if AddControllers is used. It is unexpected to be added by AddControllers method.

There should be **AddControllersCore** similar toAddMvcCore which allows to AddControllers without adding authentication and authorization

**Comment by BearsPunch** [comment]
> > This was surprising to me as well, why does framework adds authorization without explicitly asking for it? AddControllers should do what it says just add the controllers, Why do I need to pay for services that I didn't ask for. I do understand that to make building applications with asp.net more approachable for beginners, you need to hide certain complexity. But this is some unexpected addition of functionality from framework itself.
> 
> Did you run into any unexpected behavior?

Sorry David, but this question is going from your team leader? If yes, ok, I can understand it, because managers mostly don't understand technic staff. 
But your question sounds like - We will not change / modify the wiring in the house until the whole house starts to burn.

**Comment by davidfowl** [comment]
> Sorry David, but this question is going from your team leader? If yes, ok, I can understand it, because managers mostly don't understand technic staff.

That gave me a good laugh, thanks.

> But your question sounds like - We will not change / modify the wiring in the house until the whole house starts to burn.

https://github.com/dotnet/aspnetcore/issues/32485 this might be the solution to your problems. Also AddControllersCore isn't a bad idea, I just wanted to understand if there were actually issues customers were running into. It seems like the answer is no.

**Comment by ctbhavius** [comment]
Respectfully, I will disagree with the statement that "it seems like the answer is no" about customers running into issues.  I write from the perspective of a software engineer who has been coding in a wide variety of programming languages over many decades. I support the position of BearsPunch that code should perform as designed and as represented.  After all, isn't that what separation of concerns is all about?  For the record, in my experience, Microsoft's implementation (and changing implementations) of identity authentication and authorization has been the NUMBER ONE cause of headaches and aggravation over the past decade or more. So I concur completely with BearsPunch that .AddControllers() should do that and only that.  Method calls to add separate services for authentication and authorization should indeed be SEPARATE so that controllers and code can be tested WITHOUT any worrying about concerns that authentication/authorization has been added to the pipeline somewhere "magically and mysteriously".  It is difficult to trust implementation/execution of a pipeline if the design of the code cannot be trusted and does not respect foundational design principles.

**Comment by davidfowl** [comment]
There are 2 mitigations:
- We added [CreateEmptyBuider](https://github.com/dotnet/aspnetcore/issues/32485), and there's no "magic" nor any features that are on by default, power users can go crazy with this and trim things down significantly.
- We can look into removing that call from MVC in a future release when we have time to asses the impact and see if it breaks anything. 

The issue isn't going to be closed, but we're not going to be spending any time looking at this in .NET 8 (opportunity costs). If you are very motivated, it would help if an investigation could do an initial assessment of removing that call and chasing down potential breaking changes).

**Comment by divyeshio** [comment]
`CreateBuilder`, `CreateSlimBuilder` and now `CreateEmptyBuilder` finally does what it should have been in the first place. If we look at other frameworks out there like express, next, django, flask. None of them adds authentication by default, it is opt in.
